feat: adding support for examples directory

dandhlee opened the pull request:
Currently, the tarball expands into `obj/api` which can be problematic when developers want to have `api` directory to contain the YAML files inside it. To avoid having `obj/api/api/*`, if there is an `api` directory detected in the tarball it will simply expand objects only into `obj`.

Fixes internally filed issue

dandhlee committed:
feat: adding support for examples directory

jskeet reviewed (approved):
I'm assuming that:

- tar.decompress doesn't cause issues if sees a directory that already exists? (If it does, you could just move the api_path.mkdir call to after the new code)
- tar files always contain directory entries for any directories they require (so you couldn't have a tar file with a single entry of "api/xyz.yml"?

dandhlee commented:
Yes. At the point in which `tar.decompress` happens, the working directory will be in a temporary directory so there shouldn't be an existing directory anyways. If it does have an existing directory (i.e. obj/api already exists and the tarball decompresses an api directory in obj/*) tar command should simply add to the existing directory. In case of a file discrepancy it will be overwritten with the tarball's content.

No, most of the tar files at the moment don't contain any directories whatsoever. If so, we'll unzip to `obj/api`, if `api` directory exists then we'll simply only unpack to `obj`.

jskeet commented:
I think you may have misunderstood my second question.

Imagine there is a tar file that contains a single file: `api/xyz.yml`. Is there any guarantee that there'll also be an entry for the *directory* `api`? The latter is what your code is testing for - if it's possible to have entries for "files inside `api`" without an entry for "the `api` directory", this detection code won't work.

dandhlee commented:
If you're asking about whether tarfile library can detect the `api` directory itself in the tarfile, yes. Even if you don't specifically include that `api` directory be a separate entry, the library will see that any files with `dir_x/file_xyz` will have a separate entry it sees for `dir_x` and that's what the code checks and sees if it is a directory also called `api`

jskeet commented:
Right. I didn't know whether it was guaranteed that there *would* be that separate entry, that's all. LGTM.

dandhlee commented:
Sounds good! Glad I was able to clear that up for you :)

tbpg commented on docpipeline/generate.py:
```diff
@@ -86,6 +87,14 @@ def process_blob(blob, credentials, devsite_template):
+    # If so, only compress things into obj/*
+    tar_file = tarfile.open(tar_filename)
+    for tarinfo in tar_file:
+        if tarinfo.isdir() and "api" in tarinfo.name:
```
Should this be `in` or `==`?

dandhlee replied:
either works, but we would have a problem if we got a directory with the name containing `api` but not an api directory we're looking for. Thanks for catching that!

tbpg commented on docpipeline/generate.py:
```diff
@@ -86,6 +87,14 @@ def process_blob(blob, credentials, devsite_template):
     blob.download_to_filename(tar_filename)
     log.info(f"Downloaded gs://{blob.bucket.name}/{blob.name} to {tar_filename}")
 
+    # Check to see if api directory exists in the tarball.
```
Will you add a test for this? We probably want another test case that does one or two of the same checks as `test_generate`. Can make a new `@pytest.fixture` that copies `testdata` into `tmpdir/api` then returns `tmpdir`.

dandhlee replied:
added! Please see.

dandhlee committed:
Merge branch 'master' of github.com:googleapis/doc-pipeline into examples

dandhlee committed:
Merge branch 'master' into examples

dandhlee committed:
Merge branch 'master' of github.com:googleapis/doc-pipeline into examples

dandhlee committed:
adding test case

dandhlee committed:
Merge branch 'examples' of github.com:googleapis/doc-pipeline into examples

dandhlee committed:
fixing testcase and modifying generate

dandhlee commented:
Added a test case which tests with a tarball that contains `api` directory, and modified `generate.py` so that it only decompresses into `obj` when necessary, without modifying `api_path` variable

tbpg commented on tests/test_apidir.py:
```diff
@@ -0,0 +1,108 @@
+    return tmpdir
+
+
+def test_apidir(api_dir, tmpdir):
```
Mind moving this into the existing test file? Looking at it now, we may be able to just use the same `api_dir` fixture -- just use `tmpdir` not `api_dir` when running `docuploader`.

If we can't use the same fixture, please update the name (`api_dir` should return the `api/` directory).

Also, there may be parts we can refactor into common helpers to avoid duplicating code.

dandhlee replied:
Fixed, merged the two test cases and refactored!

dandhlee committed:
adding a refactored test case

tbpg commented on tests/test_generate.py:
```diff
@@ -47,17 +55,30 @@ def test_generate(yaml_dir, tmpdir):
+
+
+# Fetches the pages used for testing
+def fetch_pagedata(storage_client, test_bucket):
```
Perhaps name this `setup_tesdata`? Not sure the best name, but `pagedata` doesn't seem accurate.

dandhlee replied:
`testdata` sounds more inclusive for the variety here, I'll go ahead with that

tbpg reviewed (changes requested):
Looking good! A few more naming/nit type comments.

dandhlee commented on tests/test_generate.py:
```diff
@@ -47,17 +55,30 @@ def test_generate(yaml_dir, tmpdir):
+
+
+# Clean up any previous test data in the bucket.
+def cleanup_data(storage_client, test_bucket, yaml_blob, html_blob, bucket, tmpdir):
```
fixed!

dandhlee commented on tests/test_generate.py:
```diff
@@ -70,7 +91,7 @@ def test_generate(yaml_dir, tmpdir):
             "--destination-prefix=docfx",
         ],
-        cwd=yaml_dir,
+        cwd=target_dir,
```
fixed!

dandhlee commented on tests/test_generate.py:
```diff
@@ -47,17 +55,30 @@ def test_generate(yaml_dir, tmpdir):
         html_blob.delete()
 
+
+# Upload the tarball and verify, then run docfx
```
confused that with `builds docs`. Fixed!

dandhlee commented on tests/test_generate.py:
```diff
@@ -112,6 +136,36 @@ def test_generate(yaml_dir, tmpdir):
+def test_apidir(api_dir, tmpdir):
+    test_bucket, credentials, storage_client = test_init()
+
+    yaml_blob_name, html_blob_name, bucket, yaml_blob, html_blob = fetch_pagedata(
```
done!

dandhlee committed:
reflecting changes

dandhlee commented:
Fixed! Please take a look again :D

tbpg reviewed (approved):
LGTM. Adding @crwilcox for Python style.

crwilcox commented on docpipeline/generate.py:
```diff
@@ -91,6 +92,7 @@ def format_docfx_json(metadata):
 def process_blob(blob, credentials, devsite_template):
     tmp_path = pathlib.Path(tempfile.TemporaryDirectory(prefix="doc-pipeline.").name)
     api_path = tmp_path.joinpath("obj/api")
+    decompress_path = api_path
```
is there a reason to not remove `api_path` as a name and replace it with `decompress_path`? I can see how that is a more understand name, and I see calls to `api_path` have been updated as well.

dandhlee replied:
`decompress_path` only gets used to determine where we should decompress the tar content into, whereas `api_path` gets used a lot more further down. I have tried just using `api_path` alone but when we have a condition where `api/*` directory is encountered in the tarball, then we'd have to update it before decompressing (from `/obj/api` to `obj`), then update it back to `obj/api` which I thought was a bit too much unnecessary changes.

Ultimately it needs to end up in `obj/api` so I only wanted to use and update a separate variable `decompress_path`.

crwilcox replied:
The bit I am 'clicking into' is that this is aliasing another variable, within the same function. My take on this is that either
1) this is unnecessary and doesn't add clarity
2) this is a new more informative name
3) this is a sign this area should be abstracted.

At minimum I would change this to be api_path=decompress_path=value

crwilcox commented on docpipeline/generate.py:
```diff
@@ -101,8 +103,16 @@ def process_blob(blob, credentials, devsite_template):
-    log.info(f"Decompressed {blob.name} in {api_path}")
+    # Check to see if api directory exists in the tarball.
+    # If so, only decompress things into obj/*
+    tar_file = tarfile.open(tar_filename)
```
I assume this is to allow a broader set of tars? (ones that are the contents of a directory vs a directory itself)

Is it possible we could have obj/api/api?

Is this only used by automation? Do we get diverse formats of tar through there?

dandhlee replied:
- yes
- without this snippet we could have `obj/api/api` if people were to include `api/*` in their tarballs, but we do not expect people to have `api/api/*` in the tarballs
- we don't get a diverse format, but we want to expand support where it makes sense, such as this case where the files already come packaged in `api/*` directory

crwilcox replied:
My concern is this expanded support would malform `api/api/*` tars right? since it would navigate inward? Are people manually creating tarsballs? I don't quite understand the motivation for expanding this if it is only used by bots. I may misunderstand how this is used.

dandhlee replied:
- It would, but the original would not have supported it either way and we do not expect people to have `api/api/*`at all.
- People are not manually creating tarballs afaik, but they're automating the process and curating how they are set up.
- Originally this support was included because all the tarballs we got just had everything laid flat out, but for `.net` Jon extracted things into `api/*` and `examples/*` and some laid flat out, which would not have been caught in the current format. So this support expands from just flat tarballs to ones that also have stuff laid in `api/*`

crwilcox replied:
👍

dandhlee committed:
combining two variable assignments
